modules/apps/client-extension/client-extension-web/src/main/resources/META-INF/resources/js/util.ts

```diff
@@ -0,0 +1,33 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+/**
+ * Dynamically load an ES module and extract a symbol from it.
+ *
+ * This function is provided so that we don't need to constantly tweak tsc and
+ * webpack to be able to load things dynamically.
+ *
+ * @param moduleName The path to the ES module
+ * @param symbol The exported symbol to return
+ */
+export async function loadClientExtension<T>(
+	moduleName: string,
+	symbol: string
+): Promise<T> {
+
+	// @ts-ignore
+	const module = await import(/* webpackIgnore: true*/ moduleName);
+
+	return module[symbol];
```

Comment:
It's not that it is cached, it's that it can only be loaded once, since it is a JS module and may maintain global state (as it is the case for Node.js modules or .so/.dll shared libraries).

In more detail: whenever you import anything there are three steps involved:

1. Download the file
2. Parse it
3. Load it into the JS engine memory space

Step 1 is affected by the HTTP cache and, once the file is downloaded, if it doesn't change, it should be cached (though I'm not 100% sure of what cache headers are we sending right now. We may need to revisit this in the near/mid future).

Step 2: I'm not sure how browsers cache the JS preprocessing. I would say it depends on the browser JS engine and its settings (if JIT is enabled, if the cache is per browser or page or site session, etc.)

Regarding step 3 it happens once per refresh: if you import something that is not loaded in the JS engine, it is loaded, executed and never loaded again (bc it would reset its state).